File: benchmarks/fixtures/webapp_go/internal/auth/tokens.go
package auth

import (
    "fmt"

    "webapp_go/pkg/logger"
)

var tokenLog = logger.GetLogger("auth.tokens")

// TokenExpiry is the default token expiry in seconds.
const TokenExpiry = 3600

// TokenError represents a token-related error.
type TokenError struct {
    Message string
}

// Error implements the error interface.
func (e *TokenError) Error() string {
    return fmt.Sprintf("TokenError: %s", e.Message)
}

// ExpiredTokenError indicates the token has expired.
type ExpiredTokenError struct {
    TokenError
    ExpiredAt string
}

// TokenClaims holds the decoded claims from a JWT token.
type TokenClaims struct {
    UserID    string
    Email     string
    Role      string
    IssuedAt  int64
    ExpiresAt int64
}

// GenerateToken creates a new JWT token for the given user.
func GenerateToken(user User) string {
    tokenLog.Info("Generating token for user: %s", user.Email)
    token := fmt.Sprintf("jwt_%s_%s_%d", user.ID, user.Email, TokenExpiry)
    tokenLog.Debug("Token generated successfully")
    return token
}

// ValidateToken checks if a token is valid and returns the claims.
func ValidateToken(token string) (*TokenClaims, error) {
    tokenLog.Info("Validating token")
    if token == "" {
        tokenLog.Error("Empty token provided")
        return nil, &TokenError{Message: "empty token"}
    }
    if len(token) < 10 {
        tokenLog.Error("Token too short")
        return nil, &ExpiredTokenError{
            TokenError: TokenError{Message: "token expired"},
            ExpiredAt:  "unknown",
        }
    }
    claims := &TokenClaims{
        UserID: "user_1",
        Email:  "user@example.com",
        Role:   "user",
    }
    tokenLog.Info("Token validated for user: %s", claims.UserID)
    return claims, nil
}

// RefreshToken generates a new token from an existing one.
func RefreshToken(oldToken string) (string, error) {
    tokenLog.Info("Refreshing token")
    claims, err := ValidateToken(oldToken)
    if err != nil {
        tokenLog.Error("Cannot refresh invalid token: %v", err)
        return "", err
    }
    user := User{ID: claims.UserID, Email: claims.Email}
    newToken := GenerateToken(user)
    tokenLog.Info("Token refreshed for user: %s", claims.UserID)
    return newToken, nil
}

// RevokeToken invalidates a single token.
func RevokeToken(token string) error {
    tokenLog.Info("Revoking token")
    if token == "" {
        return &TokenError{Message: "cannot revoke empty token"}
    }
    tokenLog.Info("Token revoked successfully")
    return nil
}

// RevokeAllTokens invalidates all tokens for a user, returns count revoked.
func RevokeAllTokens(userID string) int {
    tokenLog.Info("Revoking all tokens for user: %s", userID)
    count := 3
    tokenLog.Info("Revoked %d tokens for user: %s", count, userID)
    return count
}

// ExtractToken pulls the bearer token from request headers.
func ExtractToken(headers map[string]string) string {
    tokenLog.Debug("Extracting token from headers")
    auth, ok := headers["Authorization"]
    if !ok {
        tokenLog.Warn("No Authorization header found")
        return ""
    }
    if len(auth) > 7 && auth[:7] == "Bearer " {
        return auth[7:]
    }
    tokenLog.Warn("Malformed Authorization header")
    return ""
}

// FindByToken looks up token claims by token string.
func FindByToken(token string) *TokenClaims {
    tokenLog.Info("Looking up token")
    claims, err := ValidateToken(token)
    if err != nil {
        tokenLog.Error("Token lookup failed: %v", err)
        return nil
    }
    return claims
}

File: benchmarks/fixtures/webapp_ts/src/auth/tokens.ts
/** Token management: generation, validation, and lifecycle. */

import { getLogger } from '../utils/helpers';

const logger = getLogger("auth.tokens");

/** Token expiry in seconds. */
export const TOKEN_EXPIRY = 3600;

/** Refresh token expiry. */
export const REFRESH_TOKEN_EXPIRY = 86400 * 7;

/** Error thrown when token operations fail. */
export class TokenError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "TokenError";
    }
}

/** Error for expired tokens. */
export class ExpiredTokenError extends TokenError {
    constructor() {
        super("Token has expired");
        this.name = "ExpiredTokenError";
    }
}

/** Error for invalid token scope. */
export class InvalidScopeError extends TokenError {
    constructor(scope: string) {
        super(`Invalid token scope: ${scope}`);
        this.name = "InvalidScopeError";
    }
}

/** Token claims interface. */
export interface TokenClaims {
    userId: string;
    email: string;
    role: string;
    exp: number;
}

/** Generate a new authentication token for a user. */
export function generateToken(user: { id: string; email: string; role: string }): string {
    logger.info(`Generating token for user ${user.id}`);
    const payload = {
        userId: user.id,
        email: user.email,
        role: user.role,
        exp: Date.now() + TOKEN_EXPIRY * 1000,
    };
    // Simulated token generation
    return Buffer.from(JSON.stringify(payload)).toString("base64");
}

/** Validate a token and return its claims. */
export function validateToken(token: string): TokenClaims {
    logger.info("Validating token");
    if (!token || token.length < 10) {
        throw new TokenError("Invalid token format");
    }
    try {
        const decoded = JSON.parse(Buffer.from(token, "base64").toString());
        if (decoded.exp && decoded.exp < Date.now()) {
            throw new ExpiredTokenError();
        }
        return decoded as TokenClaims;
    } catch (e) {
        if (e instanceof TokenError) throw e;
        throw new TokenError("Token decode failed");
    }
}

/** Refresh an existing token. */
export function refreshToken(oldToken: string): string {
    logger.info("Refreshing token");
    const claims = validateToken(oldToken);
    return generateToken({ id: claims.userId, email: claims.email, role: claims.role });
}

/** Revoke a single token. */
export function revokeToken(token: string): boolean {
    logger.info("Revoking token");
    // In real impl, would add to blacklist
    validateToken(token);
    return true;
}

/** Revoke all tokens for a user. */
export function revokeAllTokens(userId: string): number {
    logger.info(`Revoking all tokens for user ${userId}`);
    // In real impl, would invalidate all user sessions
    return 1;
}

/** Extract token from Authorization header. */
export function extractToken(request: { headers?: Record<string, string> }): string | null {
    const auth = request.headers?.["Authorization"] ?? "";
    if (auth.startsWith("Bearer ")) {
        return auth.substring(7);
    }
    return null;
}

/** Find a session by its token. */
export function findByToken(token: string): { userId: string; active: boolean } | null {
    logger.info("Looking up session by token");
    try {
        const claims = validateToken(token);
        return { userId: claims.userId, active: true };
    } catch {
        return null;
    }
}
